
<h1 align="center"> Battle of the Neighbourhoods - Toronto </h1>

In [1]:
import pandas as pd

# Question 1

## Importing Data

In [2]:
import requests

In [3]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

Response 200 means that we are able to make the connection to the page

In [5]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [6]:
len(wiki_data), type(wiki_data)

(3, list)

We need the first table alone, so dropping the other tables

In [7]:
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

Dropping Borough which are not assigned

In [8]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Grouping the records based on Postal Code

In [9]:
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Checking for number of records where Neighbourhood is "Not assigned"

In [10]:
df.Neighbourhood.str.count("Not assigned").sum()

0

In [11]:
df = df.reset_index()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

In [12]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [13]:
df.shape

(103, 3)

Answer to Question 1: We have 103 rows and 3 columns

# Question 2

Installing geocoder

In [14]:
pip install geocoder

In [15]:
import geocoder # import geocoder

Tried the below approach, ran for 20 mins, then killed it. Changing the code cell to Text for now so that the run all execution doesn't stop.

```python
# initialize your variable to None
lat_lng_coords = None

postal_code = 'M3A'

# loop until you get the coordinates
while(lat_lng_coords is None):
  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
  lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]
```

Alternatively, as suggested in the assignment, Importing the CSV file from the URL

In [16]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [17]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", data.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


Since the dimensions are the same, we can try to join on the postal codes to get the required data.

Checking the column types of both the dataframes, especially Postal Code column since we are trying to join on it

In [18]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [19]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [20]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [21]:
combined_data.shape

(103, 5)

**Solution:** We get 103 rows as expected when we do a inner join, so we have good data.

# Question 3

Drawing inspiration from the previous lab where we cluster the neighbourhood of NYC, We cluster Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [27]:
pip install geopy

Note: you may need to restart the kernel to use updated packages.


In [28]:
from geopy.geocoders import Nominatim 

In [29]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


Let's visualize the map of Toronto

In [32]:
pip install folium

Note: you may need to restart the kernel to use updated packages.


In [33]:
import folium

In [60]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

Initializing Foursquare API credentials

In [34]:
CLIENT_ID = 'J1I3PSJEAIC0AC3JFTDXG1OFYFHDQ4M534IIY0VLBNUNRLQQ' 
CLIENT_SECRET = '3EYA5HIN2XX5O3IOQHW2J2B2UHAJU3VLJMJB22A4ZKZSAFHI'
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: J1I3PSJEAIC0AC3JFTDXG1OFYFHDQ4M534IIY0VLBNUNRLQQ
CLIENT_SECRET:3EYA5HIN2XX5O3IOQHW2J2B2UHAJU3VLJMJB22A4ZKZSAFHI


Next, we create a function to get all the venue categories in Toronto

In [35]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Collecting the venues in Toronto for each Neighbourhood

In [36]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [37]:
venues_in_toronto.shape

(1321, 5)

So we have 1317 records and 5 columns. Checking sample data


In [38]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2  Victoria Village               43.725882               -79.315572   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1           Variety Store      Food & Drink Shop  
2  Victoria Village Arena           Hockey Arena  
3               Portugril  Portuguese Restaurant  
4             Tim Hortons            Coffee Shop

Checking the Venues based on Neighbourhood

In [39]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                      Victoria Village   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1307  Mimico NW, The Queensway West, South of Bloor,...   
1308  Mimico NW, The Queensway West, South of Bloor,...   
1309  Mimico NW, The Queensway West, South of Bloor,...   
1310  Mimico NW, The Queensway West, South of Bloor,...   
1311  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656           Variety Store   
2                  43.725882               -79.315572  Victoria Village Arena   
3                  43.725882               -79.315572               Portugril   
4                  43.725882               -79.315572             Tim Hortons   
...                      ...                      ...                     ...   
1307               43.628841               -79.520999        South St. Burger   
1308               43.628841               -79.520999              Wingporium   
1309               43.628841               -79.520999               Dollarama   
1310               43.628841               -79.520999          Healthy Planet   
1311               43.628841               -79.520999                7-Eleven   

             Venue Category  
0                      Park  
1         Food & Drink Shop  
2              Hockey Arena  
3     Portuguese Restaurant  
4               Coffee Shop  
...                     ...  
1307           Burger Joint  
1308            Wings Joint  
1309         Discount Store  
1310        Supplement Shop  
1311      Convenience Store  

[412 rows x 5 columns]

So there are 405 records for each neighbourhood.

Checking for the maximum venue categories

In [40]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                    Lawrence Manor, Lawrence Heights   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge      CN Tower, King and Spadina, Railway Lands, Har...   
Airport Service     CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                                     Little Portugal, Trinity   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                                     Caledonia-Fairbanks   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.718518               -79.464763   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Lounge                   43.628947               -79.394420   
Airport Service                  43.628947               -79.394420   
...                                    ...                      ...   
Warehouse Store                  43.705369               -79.349372   
Wine Bar                         43.653206               -79.400049   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.689026               -79.453512   
Yoga Studio                      43.715383               -79.321558   

                                               Venue  
Venue Category                                        
Accessories Store                Ardene Shoes Outlet  
Airport              Toronto Downsview Airport (YZD)  
Airport Food Court                 Billy Bishop Café  
Airport Lounge                         Porter Lounge  
Airport Service     Porter Airlines Check-In Counter  
...                                              ...  
Warehouse Store                               Costco  
Wine Bar                             Paris Paris Bar  
Wings Joint                               Wingporium  
Women's Store                          Maximum Woman  
Yoga Studio                                YogaSpace  

[230 rows x 4 columns]

There are around 232 different types of Venue Categories. Interesting!

## One Hot encoding the venue Categories

In [41]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Airport  Airport Food Court  Airport Lounge  \
0                     0        0                   0               0   
1                     0        0                   0               0   
2                     0        0                   0               0   
3                     0        0                   0               0   
4                     0        0                   0               0   
...                 ...      ...                 ...             ...   
1316                  0        0                   0               0   
1317                  0        0                   0               0   
1318                  0        0                   0               0   
1319                  0        0                   0               0   
1320                  0        0                   0               0   

      Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                   0                 0                    0             0   
1                   0                 0                    0             0   
2                   0                 0                    0             0   
3                   0                 0                    0             0   
4                   0                 0                    0             0   
...               ...               ...                  ...           ...   
1316                0                 0                    0             0   
1317                0                 0                    0             0   
1318                0                 0                    0             0   
1319                0                 0                    0             0   
1320                0                 0                    0             0   

      Aquarium  Art Gallery  ...  Trail  Train Station  \
0            0            0  ...      0              0   
1            0            0  ...      0              0   
2            0            0  ...      0              0   
3            0            0  ...      0              0   
4            0            0  ...      0              0   
...        ...          ...  ...    ...            ...   
1316         0            0  ...      0              0   
1317         0            0  ...      0              0   
1318         0            0  ...      0              0   
1319         0            0  ...      0              0   
1320         0            0  ...      0              0   

      Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                                 0                 0                      0   
1                                 0                 0                      0   
2                                 0                 0                      0   
3                                 0                 0                      0   
4                                 0                 0                      0   
...                             ...               ...                    ...   
1316                              0                 0                      0   
1317                              0                 0                      0   
1318                              0                 0                      0   
1319                              0                 0                      0   
1320                              0                 0                      0   

      Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                   0         0            0              0            0  
1                   0         0            0              0            0  
2                   0         0            0              0            0  
3                   0         0            0              0            0  
4                   0         0            0              0            0  
...               ...       ...          ...            ...          ...  
1316                0

Adding the neighbourhood to the encoded dataframe

In [42]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2  Victoria Village                  0        0                   0   
3  Victoria Village                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  ...  Trail  Train Station  \
0             0         0  ...      0              0   
1             0         0  ...      0              0   
2             0         0  ...      0              0   
3             0         0  ...      0              0   
4             0         0  ...      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                0         0            0              0            0  
1                0         0            0              0            0  
2                0         0            0              0            0  
3                0         0            0              0            0  
4                0         0            0              0            0  

[5 rows x 231 columns]

We will group the Neighbourhoods, calculate the mean venue categories in each Neighbourhood 

In [43]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Airport  Airport Food Court  Airport Lounge  Airport Service  \
0      0.0                 0.0             0.0              0.0   
1      0.0                 0.0             0.0              0.0   
2      0.0                 0.0             0.0              0.0   
3      0.0                 0.0             0.0              0.0   
4      0.0                 0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  ...  Trail  \
0               0.0             0.000000           0.0       0.0  ...    0.0   
1               0.0             0.000000           0.0       0.0  ...    0.0   
2               0.0             0.000000           0.0       0.0  ...    0.0   
3               0.0             0.000000           0.0       0.0  ...    0.0   
4               0.0             0.041667           0.0       0.0  ...    0.0   

   Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0            0.0                            0.0               0.0   
1            0.0                            0.0               0.0   
2            0.0                            0.0               0.0   
3            0.0                            0.0               0.0   
4            0.0                            0.0               0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 231 columns]

Let's make a function to get the top most common venue categories

In [44]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [45]:
import numpy as np

There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods

In [46]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East           Coffee Shop   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant          Skating Rink        Breakfast Spot   
1               Dance Studio                   Gym              Pharmacy   
2                Coffee Shop      Sushi Restaurant           Supermarket   
3                       Bank    Chinese Restaurant   Japanese Restaurant   
4         Italian Restaurant        Sandwich Place            Hobby Shop   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0           Yoga Studio     Electronics Store  Eastern European Restaurant   
1           Coffee Shop                   Pub               Sandwich Place   
2         Shopping Mall        Sandwich Place                  Bridal Shop   
3           Yoga Studio     Electronics Store  Eastern European Restaurant   
4     Indian Restaurant  Fast Food Restaurant             Greek Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0             Drugstore            Donut Shop                Dog Run  
1    Dim Sum Restaurant         Deli / Bodega       Department Store  
2            Restaurant           Pizza Place               Pharmacy  
3             Drugstore            Donut Shop                Dog Run  
4             Juice Bar          Liquor Store       Sushi Restaurant

Let's make the model to cluster our Neighbourhoods

In [47]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [48]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

Checking the labelling of our model

In [49]:
kmeans.labels_[0:100]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 3, 0, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4, 1,
       0, 1, 1, 1, 1, 1, 0, 0])

Let's add the clustering Label column to the top 10 common venue categories

In [50]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Join toronto_grouped with combined_data on neighbourhood to add latitude & longitude for each neighborhood to prepare it for plotting

In [51]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             0.0                  Park   
1  43.725882 -79.315572             1.0           Pizza Place   
2  43.654260 -79.360636             1.0           Coffee Shop   
3  43.718518 -79.464763             1.0        Clothing Store   
4  43.662301 -79.389494             1.0           Coffee Shop   

    2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0       Food & Drink Shop           Yoga Studio          Deli / Bodega   
1             Coffee Shop          Hockey Arena  Portuguese Restaurant   
2                    Park                Bakery         Breakfast Spot   
3  Furniture / Home Store     Accessories Store  Vietnamese Restaurant   
4        Sushi Restaurant           Yoga Studio                    Bar   

  5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
0     Electronics Store  Eastern European Restaurant             Drugstore   
1           Yoga Studio               Discount Store      Department Store   
2                  Café                      Theater            Restaurant   
3           Coffee Shop           Miscellaneous Shop              Boutique   
4              Beer Bar                Smoothie Shop        Sandwich Place   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Donut Shop               Dog Run    Distribution Center  
1          Dessert Shop    Dim Sum Restaurant                  Diner  
2                   Pub        Chocolate Shop            Yoga Studio  
3           Event Space            Donut Shop       Department Store  
4          Burger Joint         Burrito Place                   Café

Drop all the NaN values to prevent data skew

In [52]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Plotting the clusters on the map

In [53]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [54]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Let's verify each of our clusters

Cluster 1

In [55]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York             0.0                  Park   
21              York             0.0                  Park   
32       Scarborough             0.0            Playground   
35         East York             0.0                  Park   
45        North York             0.0                  Park   
64              York             0.0                  Park   
66        North York             0.0                  Park   
85       Scarborough             0.0                  Park   
91  Downtown Toronto             0.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0      Food & Drink Shop           Yoga Studio         Deli / Bodega   
21                   Bar         Women's Store           Golf Course   
32           Yoga Studio          Dance Studio     Electronics Store   
35     Convenience Store          Intersection           Yoga Studio   
45           Yoga Studio         Deli / Bodega     Electronics Store   
64     Convenience Store         Jewelry Store           Yoga Studio   
66     Electronics Store     Convenience Store           Yoga Studio   
85            Playground          Intersection          Dance Studio   
91            Playground                 Trail           Curling Ice   

          5th Most Common Venue        6th Most Common Venue  \
0             Electronics Store  Eastern European Restaurant   
21                Deli / Bodega            Electronics Store   
32  Eastern European Restaurant                    Drugstore   
35             Department Store            Electronics Store   
45  Eastern European Restaurant                    Drugstore   
64             Department Store            Electronics Store   
66                Deli / Bodega  Eastern European Restaurant   
85  Eastern European Restaurant                    Drugstore   
91  Eastern European Restaurant                    Drugstore   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                     Drugstore            Donut Shop               Dog Run   
21  Eastern European Restaurant             Drugstore            Donut Shop   
32                   Donut Shop               Dog Run   Distribution Center   
35  Eastern European Restaurant             Drugstore            Donut Shop   
45                   Donut Shop               Dog Run   Distribution Center   
64  Eastern European Restaurant             Drugstore            Donut Shop   
66                    Drugstore            Donut Shop               Dog Run   
85                   Donut Shop               Dog Run   Distribution Center   
91                   Donut Shop               Dog Run   Distribution Center   

   10th Most Common Venue  
0     Distribution Center  
21                Dog Run  
32         Discount Store  
35                Dog Run  
45         Discount Store  
64                Dog Run  
66    Distribution Center  
85         Discount Store  
91         Discount Store

Cluster 2

In [56]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1          North York             1.0           Pizza Place   
2    Downtown Toronto             1.0           Coffee Shop   
3          North York             1.0        Clothing Store   
4    Downtown Toronto             1.0           Coffee Shop   
7          North York             1.0                   Gym   
..                ...             ...                   ...   
98          Etobicoke             1.0            Smoke Shop   
99   Downtown Toronto             1.0      Sushi Restaurant   
100      East Toronto             1.0           Yoga Studio   
101         Etobicoke             1.0         Deli / Bodega   
102         Etobicoke             1.0         Grocery Store   

      2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
1               Coffee Shop          Hockey Arena  Portuguese Restaurant   
2                      Park                Bakery         Breakfast Spot   
3    Furniture / Home Store     Accessories Store  Vietnamese Restaurant   
4          Sushi Restaurant           Yoga Studio                    Bar   
7                Beer Store   Japanese Restaurant             Restaurant   
..                      ...                   ...                    ...   
98                    River           Yoga Studio          Deli / Bodega   
99       Salon / Barbershop   Japanese Restaurant   Ethiopian Restaurant   
100           Auto Workshop                  Park             Comic Shop   
101          Baseball Field      Department Store            Escape Room   
102         Supplement Shop        Sandwich Place             Kids Store   

           5th Most Common Venue        6th Most Common Venue  \
1                    Yoga Studio               Discount Store   
2                           Café                      Theater   
3                    Coffee Shop           Miscellaneous Shop   
4                       Beer Bar                Smoothie Shop   
7                    Coffee Shop               Clothing Store   
..                           ...                          ...   
98   Eastern European Restaurant                    Drugstore   
99                           Pub                  Escape Room   
100                  Pizza Place                   Restaurant   
101            Electronics Store  Eastern European Restaurant   
102         Fast Food Restaurant               Discount Store   

    7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
1        Department Store          Dessert Shop    Dim Sum Restaurant   
2              Restaurant                   Pub        Chocolate Shop   
3                Boutique           Event Space            Donut Shop   
4          Sandwich Place          Burger Joint         Burrito Place   
7                    Café    Dim Sum Restaurant        Discount Store   
..                    ...                   ...                   ...   
98             Donut Shop               Dog Run   Distribution Center   
99      Indian Restaurant        Ice Cream Shop                  Café   
100         Burrito Place               Brewery            Skate Park   
101             Drugstore            Donut Shop               Dog Run   
102          Burger Joint         Burrito Place                Bakery   

    10th Most Common Venue  
1                    Diner  
2              Yoga Studio  
3         Department Store  
4                     Café  
7       Chinese Restaurant  
..                     ...  
98          Discount Store  
99            Dessert Shop  
100         Farmers Market  
101    Distribution Center  
102          Tanning Salon  

[88 rows x 12 columns]

Cluster 3

In [57]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
83  Central Toronto             2.0          Tennis Court   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
83            Restaurant  Ethiopian Restaurant     Electronics Store   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
83  Eastern European Restaurant             Drugstore            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
83               Dog Run   Distribution Center         Discount Store

Cluster 4

In [58]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6  Scarborough             3.0  Fast Food Restaurant           Yoga Studio   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
6      Department Store           Escape Room     Electronics Store   

         6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6  Eastern European Restaurant             Drugstore            Donut Shop   

  9th Most Common Venue 10th Most Common Venue  
6               Dog Run    Distribution Center

Cluster 5

In [59]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11  Etobicoke             4.0                Bakery           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
11      Department Store           Escape Room     Electronics Store   

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11  Eastern European Restaurant             Drugstore            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
11               Dog Run    Distribution Center

We have successfully cluster Toronto neighbourhood based on venue categories!